### PR TITLE
PICARD-2757: Add a command-line option to enable audit

### PR DESCRIPTION
**Pull request by zas**
# Summary

* This is a…
  * [ ] Bug fix
  * [x] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

* JIRA ticket (_optional_): PICARD-2757

# Solution

Using Python 3.8 sys.addaudithook()

That's a debugging feature.

Try '--audit os' or '--audit all'

Typical output would be something like:

```
audit:all:158721:0.24102306365966797 os.mkdir args=(b'/home/zas/.config/MusicBrainz', 511, -1)
audit:all:158721:0.24107575416564941 open args=(b'/home/zas/.config/MusicBrainz/.Picard.ini.synclock', 'a', 525377)
audit:all:158721:0.24116802215576172 fcntl.lockf args=(32, 6, None, None, 0)
audit:all:158721:0.2515387535095215 fcntl.lockf args=(32, 8, None, None, 0)
```

Format is audit:threadid:seconds_since_start event args

# Action

### CODE REVIEW

**Comment by rdswift** [comment]
Looks like it could be a useful debugging tool.

Just a "heads up" that once it's merged we'll need to update the documentation at https://picard-docs.musicbrainz.org/en/appendices/command_line.html

**Comment by Sophist-UK** [comment]
@zas For those of us onlookers, can you give one or two use cases showing how this might be used e.g. when someone says in the forums that they are having problems saving files and / or by Picard or Picard plugin developers?

**Comment by phw** [comment]
@Sophist-UK One example zas showed me yesterday was to audit all calls to functions in the os module:

```
(.venv) ➜  picard git:(python_audit) ✗ python tagger.py -a os  
{('os',)}
audit: os.listdir with args=('/home/zas/.config/MusicBrainz/Picard/plugins',)
audit: os.listdir with args=('/home/zas/.config/MusicBrainz/Picard/plugins',)
audit: os.listdir with args=('/home/zas/src/picard/picard/plugins',)
I: 20:14:16,178 browser/browser.start:121: Starting the browser integration (127.0.0.1:8000)
audit: os.mkdir with args=(b'/home/zas/.config/MusicBrainz', 511, -1)
audit: os.mkdir with args=(b'/home/zas/.config/MusicBrainz', 511, -1)
I: 20:14:19,110 browser/browser.stop:134: Stopping the browser integration
SystemExit
```

Or here is a run tracking all calls to `open`:

```
% python tagger.py --audit=open 
QSocketNotifier: Can only be used with threads started with QThread
audit: open with args=('/run/user/1000/Picard_v2.10.0.dev1_main_pipe_file', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/locale/de/LC_MESSAGES/picard.mo', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/locale/de/LC_MESSAGES/picard-attributes.mo', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/locale/de/LC_MESSAGES/picard-constants.mo', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/locale/de/LC_MESSAGES/picard-countries.mo', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/acousticbrainz.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/acousticbrainz.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/acousticbrainz.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/acousticbrainz.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/acousticbrainz.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/additional_artists_variables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/additional_artists_variables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/additional_artists_variables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartist_website.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartist_website.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartist_website.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartistextension.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartistextension.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumartistextension.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumfoldercover/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumfoldercover/__pycache__/gio.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/albumfoldercover/__pycache__/processor.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/decade.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/decade.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/decade.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/deezerart/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/deezerart/deezer/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/deezerart/deezer/__pycache__/client.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/deezerart/deezer/__pycache__/obj.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/deezerart/__pycache__/options.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/fanarttv/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/fanarttv/__pycache__/ui_options_fanarttv.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/format_performer_tags.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/format_performer_tags.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/format_performer_tags.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/format_performer_tags.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/format_performer_tags.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/happidev_lyrics.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/happidev_lyrics.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/happidev_lyrics.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/instruments.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/instruments.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/instruments.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/loadasnat.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/loadasnat.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/loadasnat.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/losslessfuncs/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/mod/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/usr/lib/python3.11/__pycache__/dataclasses.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/opencc/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/.venv/lib/python3.11/site-packages/opencc/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/.venv/lib/python3.11/site-packages/opencc/clib/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/papercdcase.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/papercdcase.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/papercdcase.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/picard-submit-to-listenbrainz/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/picard-submit-to-listenbrainz/__pycache__/ui_config.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/picard-submit-to-listenbrainz/__pycache__/ui_submit.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/picard-submit-to-listenbrainz/__pycache__/artist_grab.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/picard-submit-to-listenbrainz/__pycache__/listenbrainz.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/remove_perfect_albums.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/remove_perfect_albums.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/remove_perfect_albums.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/replaygain2/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/replaygain2/__pycache__/ui_options_replaygain2.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/__pycache__/rperformer.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/search_engine_lookup.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/__pycache__/smart_title_case.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/submit_isrc.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/submit_isrc.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/submit_isrc.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/theaudiodb.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/theaudiodb.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/theaudiodb.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/theaudiodb.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/theaudiodb.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/tracks2clipboard.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/tracks2clipboard.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/tracks2clipboard.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/viewvariables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/viewvariables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/viewvariables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/viewvariables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/viewvariables.zip', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/workandmovement/__pycache__/__init__.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/.config/MusicBrainz/Picard/plugins/workandmovement/__pycache__/roman.cpython-311.pyc', 'r', 524288)
audit: open with args=('/home/phw/devel/musicbrainz/picard/picard/ui/__pycache__/playertoolbar.cpython-311.pyc', 'r', 524288)
audit: open with args=('/proc/sys/dev/cdrom/info', 'r', 524288)
audit: open with args=('/usr/lib/python3.11/encodings/__pycache__/idna.cpython-311.pyc', 'r', 524288)
audit: open with args=('/usr/lib/python3.11/__pycache__/stringprep.cpython-311.pyc', 'r', 524288)
I: 10:55:18,913 browser/browser.start:121: Starting the browser integration (0.0.0.0:8002)
audit: open with args=(b'/home/phw/.config/MusicBrainz/.Picard.ini.synclock', 'a', 525377)
audit: open with args=(b'/home/phw/.config/MusicBrainz/.Picard.ini.synclock', 'a', 525377)
audit: open with args=('/run/user/1000/Picard_v2.10.0.dev1_main_pipe_file', 'w', 524865)
I: 10:55:21,610 browser/browser.stop:134: Stopping the browser integration
```

Why is it opening the plugin ZIPs so many times? That might be something to investigate.

File: picard/tagger.py [hunk 1]
```diff
@@ -256,6 +256,33 @@ def __init__(self, picard_args, localedir, autoupdate, pipe_handler=None):
         if picard_args.debug or "PICARD_DEBUG" in os.environ:
             self.set_log_level(logging.DEBUG)
 
+        if sys.version_info[:3] > (3, 8):
```
Comment (phw) [inline picard/tagger.py:259]: As @zas and I discussed yesterday we will drop Python 3.7 support after Picard 2.10.

**Comment by zas** [comment]
I improved (and fixed the code) a bit.

The matched prefix is now added to the line.

If you have `os,os.mkdir` as audit command-line option parameter it will give something like:

```
audit:os: os.listdir args=('/home/zas/.config/MusicBrainz/Picard/plugins',)
audit:os: os.listdir args=('/home/zas/.config/MusicBrainz/Picard/plugins',)
audit:os: os.listdir args=('/home/zas/src/picard/picard/plugins',)
audit:os.mkdir: os.mkdir args=(b'/home/zas/.config/MusicBrainz', 511, -1)
```

It also handles correctly `os,all`, `all` being a special keyword in this context (it matches all events).
I also dropped `  with args=` and just print  `args=`.

**Review by phw — APPROVED** [review]
Code looks good.

This can give use a useful tool for gathering information on otherwise hard to track down issues. If unused this has no impact and the code is small enough to be no burden.

**Comment by zas** [comment]
@phw I did some cleanup changes, can you review again? Thanks.

File: picard/tagger.py [hunk 2]
```diff
@@ -1437,6 +1441,10 @@ def process_picard_args():
     parser.add_argument("-display", nargs=1, help=argparse.SUPPRESS)
 
     # Picard specific arguments
+    parser.add_argument("-a", "--audit", action='store',
```
Comment (phw) [inline picard/tagger.py:1444]: Should we check if `sys.addaudithook` is available and skip offering this command line option if it isn't?

**Review by phw — COMMENTED** [review]
Looks good. Just minor comments / questions, otherwise LGTM

File: picard/audit.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        # prebuild the dict, constant
+        PREFIXES_DICT = make_prefixes_dict(prefixes_string)
+
+        def event_match(event):
+            return is_matching_a_prefix(event, PREFIXES_DICT)
+
+    start_time = time.time()
+
+    def audit(event, args):
+        # we can't use log here, as it generates events
```
[the remark this answers is not in the document]
Comment (zas) [reply picard/audit.py]: Done in 255ec01cd8caa57677c0af9968

**Review by phw — APPROVED** [review]
LGTM

**Pull request merged**